courseworks/week7/xuanzhaogao/hw7.jl

```diff
@@ -0,0 +1,1137 @@
+### A Pluto.jl notebook ###
+# v0.19.22
+
+using Markdown
+using InteractiveUtils
+
+# ╔═╡ 32ef4500-d88b-46c9-aa2c-b038768c4bd3
+using LinearAlgebra
+
+# ╔═╡ 54e1b7ea-70b2-44ea-aea4-40368b10f220
+using Plots
+
+# ╔═╡ f7248907-ac01-4ede-b0be-87fe7454794a
+using ForwardDiff
+
+# ╔═╡ 1ac79634-2483-4ed8-b474-66971b29273e
+md"""
+A unit-disk embedding of a graph is a way to represent the vertices and edges of a graph in the Euclidean plane such that each vertex is mapped to a distinct point and each edge is represented by a curve that connects the endpoints of the edge. In a unit-disk embedding, each vertex of the graph is represented by a disk of unit radius and the disks corresponding to adjacent vertices intersect if and only if the corresponding vertices are adjacent in the graph. The goal of a unit-disk embedding is to find a geometric representation of the graph that preserves its connectivity properties and is aesthetically pleasing. Unit-disk embeddings have applications in wireless sensor networks, where the nodes are represented by disks and the communication range between nodes is limited to the radius of their disks.
+"""
+
+# ╔═╡ 8f183ba7-fba6-4eea-b6d3-732e7cbe554a
+N = 10
+
+# ╔═╡ adc3c8b0-f385-461e-827d-5d7ce3e7c34c
+E = [(1, 2), (1, 3),
+	(2, 3), (2, 4), (2, 5), (2, 6),
+	(3, 5), (3, 6), (3, 7),
+	(4, 5), (4, 8),
+	(5, 6), (5, 8), (5, 9),
+	(6, 7), (6, 8), (6, 9),
+	(7,9), (8, 9), (8, 10), (9, 10)]
+
+# ╔═╡ 492e08b2-bcf9-4506-91d3-39620a9f9af3
+function distance!(x, i, j)
+	return sqrt((x[2 * i - 1] - x[2 * j - 1])^2 + (x[2 * i] - x[2 * j])^2)
+end
+
+# ╔═╡ 742dba9c-2c54-4e31-a528-567690298d48
+function is_unit_disk!(x::AbstractArray)
+	E = [(1, 2), (1, 3), (2, 3), (2, 4), (2, 5), (2, 6), (3, 5), (3, 6), (3, 7),(4, 5), (4, 8), (5, 6), (5, 8), (5, 9), (6, 7), (6, 8), (6, 9), (7,9), (8, 9), (8, 10), (9, 10)]
+
+	Loss = 0
+
+	for i in 1:10
+		for j in i + 1:10
+			if (i, j) in E
+				d_ij = distance!(x, i, j)
+				if d_ij >= 1
+					Loss += 1
+				end
+			else
+				d_ij = distance!(x, i, j)
+				if d_ij <= 1
+					Loss += 1
+				end
+			end
+		end
+	end
+
+	if Loss == 0
+		println("The result is a unit disk.")
+	else
+		println("The result is not a unit disk.")
+	end
+	return Loss
+end
+
+# ╔═╡ 9d1e7408-5524-4870-9229-08d8827f4eff
+md"""
+Here we define the loss function based on distance, given as
+```math
+\begin{align}
+L &= L_{Edge} + L_{NonEdge}\\
+L_{Edge} &= \sum_{i, j} 2 * {|x - 1|}^{0.2} * \text{sign}(x - 1)\\
+L_{NonEdge} &= - \sum_{i, j} {|x - 1|}^{0.2} * \text{sign}(x - 1)
```

Comment:
The threshold should be slightly larger/smaller than 1.

Reply:
The loss functions have be revised and the cutoffs are changed as 0.95 and 1.05.